internal/beef/beef.go

```diff
@@ -0,0 +1,277 @@
+package beef
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/libsv/go-bt/v2"
+)
+
+const (
+	versionBytesCount = 2
+	markerBytesCount  = 2
+	hashBytesCount    = 32
+	maxTreeHeight     = 64
+)
+
+const (
+	BEEFMarkerPart1 = 0xBE
+	BEEFMarkerPart2 = 0xEF
+)
+
+const (
+	HasNoBump = 0x00
+	HasBump   = 0x01
+)
+
+type TxData struct {
+	Transaction *bt.Tx
+	BumpIndex   *bt.VarInt
+
+	txID string
+}
+
+type BEEF struct {
+	BUMPs        BUMPs
+	Transactions []*TxData
+}
+
+func CheckBeefFormat(txHex []byte) bool {
+	if len(txHex) < versionBytesCount+markerBytesCount {
+		return false
+	}
+
+	// removes version bytes
+	txHex = txHex[versionBytesCount:]
+
+	if txHex[0] != BEEFMarkerPart1 || txHex[1] != BEEFMarkerPart2 {
+		return false
+	}
+
+	return true
+}
+
+func DecodeBEEF(beefHex []byte) (*bt.Tx, *BEEF, []byte, error) {
+	beefBytes, err := extractBytesWithoutVersionAndMarker(beefHex)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	bumps, remainingBytes, err := decodeBUMPs(beefBytes)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	transactions, remainingBytes, err := decodeTransactionsWithPathIndexes(remainingBytes)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	decodedBeef := &BEEF{
+		BUMPs:        bumps,
+		Transactions: transactions,
+	}
+
+	return decodedBeef.GetLatestTx(), decodedBeef, remainingBytes, nil
+}
+
+func (d *BEEF) GetLatestTx() *bt.Tx {
+	return d.Transactions[len(d.Transactions)-1].Transaction // get the last transaction as the processed transaction - it should be the last one because of khan's ordering
+}
+
+func decodeBUMPs(beefBytes []byte) ([]*BUMP, []byte, error) {
```

Comment:
I would try to use bump (and its decoding) implementation from `go-bc` repository

Reply:
Done